### PR TITLE
update installation strategy

### PR DESCRIPTION
**Pull request by anthonyyangdev**
This avoids the assumption of what `pip` is named as and more.

Change summary:
- Remove dependency of `harmonylang.commandPath` and `harmonylang.libraryPath` (this one isn't even being used)
- Install `harmony-model-checker` by calling pip as a Python module:
  - `/path/to/python3 -m pip install --upgrade --user harmony_model_checker`
- Invoke `harmony` by calling from the Python interpreter instead of calling the `harmony` script file:
  - `/path/to/python3 -c 'import re\nimport sys\nfrom harmony_model_checker.main import main\nif __name__ == \'__main__\':\n\tsys.argv[0] = re.sub(r\'(-script\\.pwd|\\.exe)?$\', \'\', sys.argv[0])\n\tsys.exit(main())' ...args`

### CODE REVIEW

**Review by kevinsun-dev — APPROVED** [review]
Added User-Readable component, review/comment before merging.

File: src/vscode-commands/install.ts [hunk 1]
```diff
@@ -34,3 +35,21 @@ export default async function runInstall() {
...
 }
+
+/**
+ * Parses the output messages from pip to produce
+ * user readable output.
+ */
+export async function printReadable(msgs:string) {
+    const msgLines = msgs.trim().split('\n');
+    const lastLine = msgLines[msgLines.length - 1];
+    if (lastLine.startsWith('Requirement already satisfied: ')){
```
Comment (anthonyyangdev) [inline src/vscode-commands/install.ts:46]: We specifically would probably want `'Requirements already satisfied: harmony-model-checker'`

File: src/vscode-commands/install.ts [hunk 2]
```diff
@@ -34,3 +35,21 @@ export default async function runInstall() {
         });
     });
 }
+
+/**
+ * Parses the output messages from pip to produce
+ * user readable output.
+ */
+export async function printReadable(msgs:string) {
```
Comment (anthonyyangdev) [inline src/vscode-commands/install.ts:43]: Rename this to something like `printReadableInstallMessage`; `printReadable` seems a bit general of a name without the documentation comment.

Comment (kevinsun-dev) [reply src/vscode-commands/install.ts]: Changed the function name

File: src/vscode-commands/install.ts [hunk 3]
```diff
@@ -1,21 +1,55 @@
...
+ * Parses the output messages from pip to produce
+ * user readable output.
+ */
+export async function printReadableInstallMessage(msgs:string) {
+    const msgLines = msgs.trim().split('\n');
+    const lastLine = msgLines[msgLines.length - 1];
+    if (lastLine.startsWith('Requirement already satisfied: ')){
+        Message.info('Harmony already installed.');
+    } else {
+        if (lastLine.startsWith('ERROR:')){
```
Comment (anthonyyangdev) [inline src/vscode-commands/install.ts:49]: Why only if it begins with `'ERROR:'`?

File: src/vscode-commands/install.ts (hunk 1 above)
Comment (kevinsun-dev) [reply src/vscode-commands/install.ts:46]: The first line often reports back about the requirement of 'harmony-model-checker' being satisfied, but if it was, the only things it reports back was 'Requirements already satisfied: ...', so I just took it as is.

File: src/vscode-commands/install.ts (hunk 3 above)
Comment (kevinsun-dev) [reply src/vscode-commands/install.ts:49]: It begins with ERROR because that last condition is "Successfully installed 'harmony-model-checker'", so this condition is just to report any errors.

File: src/vscode-commands/install.ts [hunk 4]
```diff
@@ -1,21 +1,55 @@
...
-        ProcessManager.startCommand(INSTALL_HARMONY_COMMAND, {}, (err, stdout, stderr) => {
+        if (!pythonPath) {
+            reject('Could not find a python path. Please install Python3 or report this if you believe it is an error.');
+            return;
+        }
+        ProcessManager.startCommand([pythonPath, '-m', 'pip', ...INSTALL_HARMONY_COMMAND_ARGS], {}, (err, stdout, stderr) => {
             OutputConsole.clear();
             if (err) {
-                reject(stdout + '\n\n' + stderr);
+                reject(stdout + '\n\n' + stderr + '\n\n');
```
Comment (anthonyyangdev) [inline src/vscode-commands/install.ts:29]: Not really an issue, but are these newlines needed?

Comment (kevinsun-dev) [reply src/vscode-commands/install.ts:29]: Huh, I don't recall adding that. Remove it or keep it, up to you.

**Comment by anthonyyangdev** [comment]
I'm gonna hold off on merging this for a bit. I want to ensure that with the new installation strategy, if it upgrades, then it should affect the harmony at the `commandPath` value in the user's `settings.json`.

File: src/vscode-commands/install.ts (hunk 3 above)
Comment (anthonyyangdev) [reply src/vscode-commands/install.ts:49]: It looks like this is not exhaustive, e.g.:
```
ERROR: Could not find a version that satisfies the requirement something-that-failed (from versions: none)
ERROR: No matching distribution found for something-that-failed
WARNING: You are using pip version 20.3.4; however, version 22.0.3 is available.
You should consider upgrading via the '/bin/python -m pip install --upgrade pip' command.
```

**Review by kevinsun-dev — APPROVED** [review]
Checks out as far as I can tell, gonna push a quick change to that output parser below.

File: src/vscode-commands/install.ts [hunk 5]
```diff
@@ -51,14 +51,28 @@ export default async function runInstall() {
...
+        // If Harmony is already installed, that's all they need to know.
         Message.info('Harmony already installed.');
     } else {
-        if (lastLine.startsWith('ERROR:')){
-            Message.error(lastLine);
-        } else {
-            Message.info(lastLine);
-        }
+        // If Harmony isn't installed, this is either a successful install or a non-pip error.
+        Message.info(msgLines[msgLines.length - 1]);
```
Comment (anthonyyangdev) [inline src/vscode-commands/install.ts]: It'd probably be helpful just output the entire error message onto the output console rather than assume just the last line contains the important information.

File: src/vscode-commands/install.ts [hunk 6]
```diff
@@ -1,21 +1,87 @@
...
+ */
+export async function printReadableInstallMessage(msgs:string) {
+    const MAX_MESSAGES = 3;
+    // Reversing the message order, as higher priority errors are ordered closer to the bottom.
+    const msgLines = msgs.trim().split('\n').reverse();
+    let highestOutputLevel = 0;
+    let harmonyInstalled = false;
+    let totalMessages = 0;
+
+    for (const mIndex in msgLines){
```
Comment (anthonyyangdev) [inline src/vscode-commands/install.ts:60]: You can use `const msg of msgLines` to iterate over the strings themselves.

File: src/vscode-commands/install.ts [hunk 7]
```diff
@@ -1,21 +1,87 @@
...
+            totalMessages++;
+            Message.error(msg);
+        } else if (msg.startsWith('WARNING:')) {
+            highestOutputLevel = 1;
+            totalMessages++;
+            Message.warn(msg);
+        }
+
+        // If too many notifications are sent, none are displayed.
+        if (totalMessages >= MAX_MESSAGES) break;
```
Comment (anthonyyangdev) [inline src/vscode-commands/install.ts:73]: Why not display the `MAX_MESSAGES` number of messages?

File: src/vscode-commands/install.ts [hunk 8]
```diff
@@ -1,21 +1,87 @@
...
+
+        if (msg.includes('Requirement already satisfied: harmony-model-checker')){
```
Comment (anthonyyangdev) [inline src/vscode-commands/install.ts:75]: Could it be possible for there to be enough `WARNINGS` only that get caught that exceed `MAX_MESSAGES` such that this branch never gets hit?

File: src/vscode-commands/install.ts (hunk 5 above)
Comment (kevinsun-dev) [reply src/vscode-commands/install.ts]: The entire error message is outputted to the console, this is only to determine how much of it makes it into those popup notifications.

File: src/vscode-commands/install.ts (hunk 6 above)
Comment (kevinsun-dev) [reply src/vscode-commands/install.ts:60]: Ooh, thanks. I was looking for that.

File: src/vscode-commands/install.ts (hunk 7 above)
Comment (kevinsun-dev) [reply src/vscode-commands/install.ts:73]: Sorry, let me rephrase. This is to prevent too many notifications from being sent. If too many are sent, none are displayed, thus we're limiting the totalMessages to 3, which are displayed.

File: src/vscode-commands/install.ts (hunk 8 above)
Comment (kevinsun-dev) [reply src/vscode-commands/install.ts:75]: Hmm, that is a distinct possibility... Let me fix that.

Comment (anthonyyangdev) [reply src/vscode-commands/install.ts:74]: looks good now
